src/main/java/spoon/reflect/visitor/chain/CtQueryImpl.java

```diff
@@ -0,0 +1,452 @@
+		FunctionWrapper(CtFunction<?, ?> code) {
+			fnc = (CtFunction<Object, Object>) code;
+		}
+
+		@SuppressWarnings("unchecked")
+		@Override
+		public void processInput(Object input) {
+			Object result;
+			try {
+				result = fnc.apply(input);
+			} catch (ClassCastException e) {
+				onClassCastException(this, e, input);
+				return;
+			}
+			if (result == null) {
+				return;
+			}
+			if (result instanceof Boolean) {
+				//the code is a predicate. send the input to output if result is true
+				if ((Boolean) result) {
+					next.accept(input);
+				} else {
+					log(this, "Skipped element, because CtFunction#accept(input) returned false", input);
+				}
+				return;
+			}
+			if (result instanceof Iterable) {
+				//send each item of Iterable to the next step
+				for (Object out : (Iterable<Object>) result) {
+					next.accept(out);
+				}
+				return;
+			}
+			if (result.getClass().isArray()) {
+				//send each item of Array to the next step
+				for (int i = 0; i < Array.getLength(result); i++) {
+					next.accept(Array.get(result, i));
+				}
+				return;
+			}
+			next.accept(result);
+		}
+	}
+
+	/**
+	 * a step which scans all children of input element and only elements matching filter go to the next step
+	 */
+	private class ChildrenFilteringWrapper extends CtScanner implements Step {
+
+		private String name;
+		protected CtConsumer<Object> next;
+		private Filter<CtElement> filter;
+
+		@SuppressWarnings("unchecked")
+		ChildrenFilteringWrapper(Filter<? extends CtElement> filter) {
+			this.filter = (Filter<CtElement>) filter;
+		}
+
+		@Override
+		public void accept(Object input) {
+			if (input == null) {
+				return;
+			}
+			log(this, "CtScanner received", input);
+			scan(input);
+		}
+		@Override
+		public void scan(CtElement element) {
+			processFilter(element);
+			super.scan(element);
+		}
+
+		private void processFilter(CtElement element) {
+			if (element == null) {
+				return;
+			}
+			boolean matches = false;
+			try {
+				matches = filter.matches(element);
+			} catch (ClassCastException e) {
+				onClassCastException(this, e, element);
+				return;
+			}
+			if (matches) {
+				//send input to output, because Fitler.matches returned true
+				next.accept(element);
+			} else {
+				log(this, "Skipped child element, because Filter#matches(input) returned false", element);
+			}
+		}
+
+		@Override
+		public CtConsumer<Object> getNext() {
+			return next;
+		}
+		@Override
+		public void setNext(CtConsumer<Object> next) {
+			this.next = next;
+		}
+		@Override
+		public String getName() {
+			return name;
+		}
+		@Override
+		public void setName(String name) {
+			this.name = name;
+		}
+		@Override
+		public boolean isFailOnCCE() {
+			//never fail on CCE during Filter.matching
+			return false;
+		}
+	}
+
+	/**
+	 * Interface used to query and manipulate elements in a functional manner.
+	 *
+	 * @param &lt;T> - the type of accepted elements
+	 */
+	private interface CtConsumer<T> {

```

Comment:
I need CtConsumer as public interface later. But Martin made it private in this first step, because after his simplifications/deleting of functionality this interface is not needed.